### PR TITLE
Support GTK keycodes

### PR DESCRIPTION
**Pull request by robert-ancell**
## Description

This PR adds support for GTK key events, which allows the Linux shell to use these instead of mapping to the GLFW event codes.

## Related Issues

First part of https://github.com/flutter/flutter/issues/57603

## Tests

Tests added to match GLFW implementation.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [ ] I wrote a migration guide: https://flutter.dev/go/breaking-changes-template *Replace this with a link to a pull request that adds the migration guide to https://flutter.dev/docs/release/breaking-changes*

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

**Comment by robert-ancell** [comment]
[Engine PR to switch to new key codes](https://github.com/flutter/engine/pull/19196).

**Review by stuartmorgan-g — APPROVED** [review]
LG, but @gspencergoog should review as well as he's more familiar with the framework side.

**Comment by stuartmorgan** [comment]
> Fixes #57603

You should mark the engine PR as fixing that issue, rather than this one, since this is the first half and doesn't fix the issue on its own.

File: dev/tools/gen_keycodes/data/keyboard_map_gtk_cc.tmpl [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include <map>
+
+// DO NOT EDIT -- DO NOT EDIT -- DO NOT EDIT
+// This file is generated by flutter/flutter@dev/tools/gen_keycodes/bin/gen_keycodes.dart and
+// should not be edited directly.
+//
+// Edit the template dev/tools/gen_keycodes/data/keyboard_maps_gtk_cxx.tmpl instead.
```
Comment (gspencergoog) [inline dev/tools/gen_keycodes/data/keyboard_map_gtk_cc.tmpl:11]: Is this the right path to edit?

File: dev/tools/gen_keycodes/lib/key_data.dart [hunk 2]
```diff
@@ -133,6 +154,13 @@ class KeyData {
   /// JSON.
   Map<String, List<String>> _nameToGlfwName;
 
+  /// The mapping from the Flutter name (e.g. "eject") to the GTK name (e.g.
+  /// "GDK_KEY_Eject").
```
Comment (gspencergoog) [inline dev/tools/gen_keycodes/lib/key_data.dart:158]: GDK?
```suggestion
  /// "GTK_KEY_Eject").
```

**Review by gspencergoog — APPROVED** [review]
32384589-a60f0e74-c078-11e7-9bc1-e5b5287aea9d

File: dev/tools/gen_keycodes/data/keyboard_map_gtk_cc.tmpl (hunk 1 above)
Comment (robert-ancell) [reply dev/tools/gen_keycodes/data/keyboard_map_gtk_cc.tmpl:11]: Yep, it's wrong. Turns out it wasn't the only case though - https://github.com/flutter/flutter/pull/60015.

File: dev/tools/gen_keycodes/lib/key_data.dart (hunk 2 above)
Comment (robert-ancell) [reply dev/tools/gen_keycodes/lib/key_data.dart:158]: GDK is the low-level part of GTK (GDK = Gimp Drawing Kit, GTK = Gimp Toolkit in original naming), so events are in the GDK namespace.

Comment (gspencergoog) [reply dev/tools/gen_keycodes/lib/key_data.dart:158]: Ahh, OK, makes sense.  I was just pattern matching...

**Pull request merged**